Use `aHash` as faster hash algorithm

Dandandan opened the pull request:
This switches the hash join to use https://github.com/rust-lang/hashbrown as hash map.

Added a small benchmark:

Before:

```
test bench_join_2_frames ... bench:   1,722,155 ns/iter (+/- 88,679)
```

After:

```
test bench_join_2_frames ... bench:   1,030,084 ns/iter (+/- 153,604)
```

ritchie46 commented:
Right.. I already was looking for a faster hashing algorithm. Do you know more about the best performant hash algorithms?

I already switched to SeaHasher in favor of FnV. Is HashBrown generally a faster default?

ritchie46 commented:
The difference is a lot btw!

ritchie46 commented:
Currently, there don't seem to be any duplicates in the bench. So the hashing may be quick, but could have more collisions. The hash algorithm is also only tested on integers. Could you also make the comparison on strings with a differing lengths? Some hash functions are more performant after a certain amount of bytes are hashed.

Dandandan commented:
I just realized hashbrown is already used by `std`, so we can switch to using `aHash` (the hash algorithm)  instead. AFAIK it's the fastest hash algorithm as of now. There are some benchmarks here https://github.com/tkaitchuck/aHash for different input sizes.

Indeed, benchmark can be improved with e.g. duplicates.

Dandandan commented:
Also added a (simple) group by test and replaced the rest of the hashmaps / hashsets to use the `aHash` hash. Also here, benches can be more complex / diversified.
Old:
```test bench_group_by      ... bench:     127,615 ns/iter (+/- 133,199)```

New
` ``test bench_group_by      ... bench:      97,031 ns/iter (+/- 61,597)```

ritchie46 commented:
I did some benchmarks locally with the code in [pandas_cmp](https://github.com/ritchie46/polars/tree/master/pandas_cmp). Here we have more realistic benches as we have duplicate keys and also test on strings.

*Lower is better*

## aHash
### join
7189.4
8180.8
8427.8

### groupby
464
3331
15063
115883

### groupby str
456
3181
29959
304080


## Seahash
### join
9099.8
9324.3
10066.9

### groupby
437
4311
30349
214479

### groupby str
623
5171
46869
458504

So yes. That are indeed some quick wins. :smiley:

Dandandan commented:
Awesome! :)

ritchie46 commented:
Hmm.. it seems that this commit broke a test. Will need to figure that out.

Dandandan commented:
Hm weird... I thought I saw a failing test before, but after it tests succeeded. Maybe something relying on ordering of hash map items?

ritchie46 commented:
> Hm weird... I thought I saw a failing test before, but after it tests succeeded. Maybe something relying on ordering of hash map items?

It was indeed the ordering. It is fixed it the upcoming commits.
